doc/bil.1
.\"    Copyright (C) 2002 Patrick Dangla
.\"
.\"    This file is part of the program BIL.
.\"
.\"    This program is free software; you can redistribute it and/or modify
.\"    it under the terms of the GNU General Public License as published by
.\"    the Free Software Foundation; either version 2 of the License, or
.\"    (at your option) any later version.
.\"
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"    GNU General Public License for more details.
.\"
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.\"    $Id: bil.1,v 1.1 2004/07/01 22:23:11 dangla Exp $
.\"
.TH Bil 1 "21 Sept 2004" "Bil 2.0" "Bil Manual Pages"
.UC 4



.SH NAME
Bil \- modeling platform based on finite element/volume methods



.SH SYNOPSIS
.B bil [option(s)] [file]



.SH DESCRIPTION
\fIBil\fR is a modeling platform based on finite element/volume methods 
dedicated to coupled problems involved in environmental engineering, 
geomechanical engineering, material sciences, etc..

.PP

If \fIfile\fR is given, \fIBil\fR compute the solution of a problem 
described in \fIfile\fR.

.PP

\fIBil\fR is developped for 1D, 2D and 3D problems. It doesn't include 
a mesh generator and post-processing treatment of outputs. However it 
can read mesh files 
created by the open-source free software \fIGmsh\fR downloadable at 
http://www.geuz.org/gmsh/.
The output files created by \fIBil\fR for 1D problems can be used easily 
by some plotting programs as \fIGnuplot\fR. But as a general rule, 
post-processing options create output files which can be used by 
\fIGmsh\fR for post-processing treatments.



.SH COMPUTATION OPTIONS

.TP 4
.B \-graph \fIfmt\fR
print the graph in \fIfile\fR\fB.graph\fR in the format \fIfmt\fR. Available formats are:
.RS
\fBmetis\fR: for METIS (if installed).
.RE
.RS
\fBhsl_mc40\fR: format given by MC40 from HSL.
.RE

.TP 4
.B \-iperm
print the inverse permutation of nodes in \fIfile\fR\fB.graph.iperm\fR by HSL_MC40.

.TP 4
.B \-eordering \fImeth\fR
print the inverse permutation of elements in file \fIfile\fR\fB.graph.iperm\fR 
by the method \fImeth\fR. Available methods are:
.RS
\fBhsl_mc43\fR: MC43 from HSL.
.RE

.TP 4
.B \-nordering \fImeth\fR
print the inverse permutation of nodes in file \fIfile\fR\fB.graph.iperm\fR 
by the method \fImeth\fR. Available methods are:
.RS
\fBhsl_mc40\fR: MC40 from HSL (same as -iperm)
.RE

.TP 4
.B \-solver \fImethod\fR
use the solver defined by \fImethod\fR. Available methods are:
.RS
\fBcrout\fR: CROUT method (default).
.RE
.RS
\fBsuperlu\fR: Sequential SuperLU (if libsuperlu.so installed).
.RE
.RS
\fBsuperlumt\fR: Multithreaded SuperLU (if libsuperlu_mt.so installed). The number of threads is specified with the option \fB-nthreads\fR (see below).
.RE
.RS
\fBsuperludist\fR: Distributed SuperLU (if libsuperlu_dist.so installed).
.RE
.RS
\fBma38\fR: MA38 from HSL (if libblas.so installed).
.RE
.RS
\fBpetscksp\fR: Petsc KSP solver (if libpetsc.so installed). Following this option we can specify the KSP method and the preconditioner method through the online commands: \fB-ksp_type <method>\fR and \fB-pc_type <method>\fR (see below).
.RE

.TP 4
.B \-nthreads \fIN\fR
use \fIN\fR threads for the solver "superlumt".

.TP 4
.B \-ff \fIN\fR
use the fill factor \fIN\fR for the multi-frontal methods SuperLU (default \fIN\fR=0 meaning that space is allocated internally by SuperLU) or MA38 (default \fIN\fR=2). This option must follow the option \fB-solver\fR \fImethod\fR.

.TP 4
.B \-ksp_type \fImethod\fR
use the KSP method \fImethod\fR for the KSP solver (default is gmres). This option must follow the option \fB-solver petscksp\fR. Available methods are:
.RS
\fBrichardson, chebyshev, cg, pipecg, groppcg, pipecgrr, cgne, fcg, pipefcg, cgls, nash, stcg, gltr, qcg, bicg, bcgs, ibcgs, qmrcgs, fbcgs, fbcgsr, bcgsl, minres, gmres, fgmres, dgmres, pgmres, pipefgmres, lgmres, cr, gcr, pipecr, pipegcr, fetidp, cgs, tfqmr, tcgmr, lsgr, symmlq, tsirm, python, tcqmr, bcgs, cgs, tfqmr, cr, lsqr, bicg, preonly\fR (see the Petsc documentation for more details at https://petsc.org).
.RE

.TP 4
.B \-pc_type \fImethod\fR
use the preconditoner method \fImethod\fR for the KSP solver (default is ilu for gmres). This option must follow the option \fB-solver petscksp\fR. Available methods are:
.RS
\fBnone, jacobi, bjacobi, sor, eisenstat, icc, ilu, asm, gasm, gamg, bddc, ksp, composite, lu, cholesky, shell\fR (see the Petsc documentation for more details at https://petsc.org).
.RE

.TP 4
.B \-with \fIinputs\fR
use the module defined in \fIinputs\fR. Available inputs are:
.RS
\fBMonolithic\fR \fIN\fR: Monolithic approach, \fIN\fR = nb of the temporal discrete solutions kept in memory that can be referred to, for possible use in time discretization schemes. \fIN\fR includes the current time at which the solution is looked for (this is the default module with \fIN\fR=2).
.RE
.RS
\fBSNIA\fR \fIN\fR: Sequential non iterative approach, \fIN\fR = nb of sequences.
.RE



.SH POST-PROCESSING OPTIONS

.TP 4
.B \-post \fIfmt\fR
generate the post-processing files \fIfile\fR\fB.pos\fR* in the format \fIfmt\fR. Available formats are: 
.RS
\fBGmshParsed\fR: for GMSH parsed file format.
.RE
.RS
\fBGmshASCII\fR : for GMSH ASCII file format.
.RE



.SH OTHER OPTIONS

.TP 4
.B \-readonly
parse \fIfile\fR, then exit. 

.TP 4
.B \-create
if \fIfile\fR doesn't exist, open it in writing mode and create an input data file by asking the user for the data.

.TP 4
.B \-debug \fIinput\fR
show data structure for \fIinput\fR. Available inputs are:
.RS
\fBgeom\fR: geometry
.RE
.RS
\fBmesh\fR: mesh
.RE
.RS
\fBmate\fR: materials
.RE
.RS
\fBfield\fR: fields
.RE
.RS
\fBinit\fR: initialization
.RE
.RS
\fBfunc\fR: functions
.RE
.RS
\fBbcond\fR: boundary conditions
.RE
.RS
\fBload\fR: loads
.RE
.RS
\fBpoin\fR: points
.RE
.RS
\fBobval\fR: objective variations
.RE
.RS
\fBiter\fR: iterative process
.RE
.RS
\fBtime\fR: time steps
.RE
.RS
\fBall\fR: all the dataset above
.RE
.RS
\fBinter\fR: interpolation functions
.RE
.RS
\fBcontinuity\fR: equations and unknowns at nodes
.RE
.RS
\fBmatrix\fR: the matrix
.RE
.RS
\fBresidu\fR: the residu
.RE
.RS
\fBnumbering\fR: matrix row/column indexes of equations/unknowns at nodes
.RE


.TP 4
.B \-help
online help.

.TP 4
.B \-info
display general informations.

.TP 4
.B \-model
display the available models.

.TP 4
.B \-model \fImymodel\fR
display an example of material properties for the model \fImymodel\fR.

.TP 4
.B \-modules
display the available modules.



.SH AUTHOR
Patrick Dangla (patrick.dangla@univ-eiffel.fr).



.SH SEE ALSO
The full documentation for Bil is maintained as a Texinfo manual. If the
\fBinfo\fR and \fBbil\fR are properly installed at your site, the command
.IP
.B info bil
.PP
should give you access to the complete manual.


CHANGELOG
News in 2.9: Use of the library "autodiff" via option at compiling time. 
Compliler c++ is forced for all source files (.c and .cpp). 
Introduction of a constitutive integrator interface for the implementation
of models.

News in 2.8: Parallelization via OpemMP and MPI.
A new algorithm is implemented in the module SNIA which is based on a 
sequential non iterative approach (SNIA). The build system is switched 
to CMake. Bugs fixed in MA38 solver. New way of using extra libraries. 
The number of temporal discrete solutions kept in memory for possible use 
in time discretization schemes can be given in the command line. 
The solver KSP from the Petsc library can be selected if the external 
library "libpetsc.so" was installed (see EXTRALIBS). 
Re-arranging some allocation of memory space for the mesh.
Reformulation of the jacobian matrix of the transformation between the
element in the reference frame and that in the current frame. Zero-thickness
element can be dealt with. Unknowns like pressure/temperature or other can be
made continuous across such zero-thickness elements. The python script
"utils/gmshcrack/gmshCrack.py" can be used to insert zero-thickness
elements in mesh file generated by GMSH.

News in 2.7: Improvements on exception mechanisms, in no case 
results will be lost.

News in 2.6: The solver of sparse unsymmetric system of linear equations MA38 
from HSL is introduced in the package and can be selected from the command line.
The solver SuperLU can also be selected if the external library "libsuperlu.so" 
was installed (see the file "make.extralibs").

News in 2.5: A new interface "Session.h" is created. Methods like
Session_Open and Session_Close can be used to open and close  sessions. 
Sessions are in a stack with the current session at the top. 
Open a session = push it, close a session = pop and delete it.
If needed instances like Message, Exception, etc... are created once in each 
session of the stack. Only those of the top session are currently used.

News in 2.4: Two new database interfaces: "HardenedCementChemistry.h" and
"CementSolutionDiffusion.h" to be used for cement-based materials. 
New models now use these interfaces. The references used to implement some
databases are now copied in the folder "src/Models/DataBases/References".

News in 2.3: Use of C++ compiler. Programmation is now extended to 
C++ language. Hence the use of .cpp and .hpp files is allowed. 
A new object "Exception.h" has been created to handle exception mechanisms 
such as interruption, floating point error.
In such occuring event the program handles the event, saves outputs and exits 
in a clean way.
Introduction of the curve builder "Expressions", based on the evaluation of 
mathematical expressions obtained from AnaGram (www.parsifalsoft.com). 
In this version a shared library, libbil.so, from the genuine sources of Bil
is created and installed in the machine. Linkage with other possible external 
libraries is possible. A new extension of input file ".conti" has been introduced 
so that we can load the solution from a previous calculation and continue this 
calculation while going through the initialization stage (see documention).

News in 2.2: WARNING, the outputs of FVM_ComputeIsotropicConductionMatrix()
and FVM_ComputeMassAndIsotropicConductionMatrix() have been modified see 
examples in models using these methods. New objects have been introduced: 
Views,  TextFile, CurvesFile, MatrixStorageFormat. Introduction of the curve 
builder "Evaluate", based on the evaluation of mathematical expressions 
obtained from Snippets (www.brokersys.com/snippets/). 
Some bugs fixed, e.g. those in "Buffer_FreeFrom". Extended field delimiters
in curves files. Some new models.

News in 2.1: A class-like structure Buffer_t has been introduced as
a circular buffer. It can be used in any functions to compute vectors or 
tables of any type. Numerical Methods, such as FEM or FVM, have been 
implemented in files such as FEM.[c,h] and FVM.[c,h]. These methods are
viewed as object with class-like structures FEM_t and FVM_t.
In these files methods have been implemented such as 
FEM_ComputeMassMatrix and can be used directly in models provided that header
has been included. The folders "Common" and "Main" have been created. 
We moved the main files in "Main" and files of common use, like "Buffer.c", 
in "Common". DataBases have been created which can be used in model files.
The matrix storage format have been implemented in a separate file 
"MatrixStorageFormat.h".

News in 2.0: The code has been rebuilt to an object-oriented programming 
code. However the code is still implemented in C language. Class-like 
structures have been introduced and implemented in separated files. 
Each class-like structure has attributes and pointers to functions 
(see e.g. structure Model_t). Moreover some improvements have been 
introduced. The file names of models (in the folder "ModelFiles") can be 
chosen abitrarily. The class-like structures "NodeSol_t" and 
"ElementSol_t" are two linked lists which contained the nodal and 
elemental solutions. A new pointer to "double" type in "ElementSol_t" 
can be used to store constant terms.

Nouveautes dans 1.8: creation d'une structure "modl_t" contenant les 
methodes (taches elementaires) i.e. des pointeurs sur fonctions.

Nouveautes dans 1.7: option "base" ajoute dans le Makefile. Qqs nouveaux
modeles. Option "Relative" possible dans OBJE, donnee apres la valeur.
Chargement automatique des modeles, il n'est plus besoin de modifier le
fichier "xmod.c". Les fonctions "dm,qm,tb,ch,in,ex,ct,mx,rs,so" sont 
renommees "dmNB,qmNB,tbNB,chNB,inNB,exNB,ctNB,mxNB,rsNB,soNB". 
Les titres des modeles ainsi que des exemples de donnees sont geres 
par les fonctions "qm1, qm2 ...". Options nouvelles de la ligne de 
commande. "bil -m" affiche les titres des modeles. "bil -m I" affiche
un exemple de donnees du modele I. Possiblite d'avoir des elements

src/Main/Bil.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>
#include <assert.h>
#include "DataSet.h"
#include "Message.h"
#include "Context.h"
#include "CommandLine.h"
#include "Help.h"
#include "Modules.h"
#include "OutputFiles.h"
#include "PosFilesForGMSH.h"
#include "Models.h"
#include "Bil.h"
#include "Exception.h"
#include "BilVersion.h"
#include "BilInfo.h"
#include "Session.h"
#include "Mry.h"
#include "TypeId.h"
#include "Threads.h"



static void   (Bil_PrintUsage)(char*) ;
static void   (Bil_PrintInfo)(void) ;
static void   (Bil_CLI)(Bil_t*) ;



Bil_t*    (Bil_Create)(int argc,char** argv)
{
  Bil_t* bil = (Bil_t*) Mry_New(Bil_t) ;
  
  Session_Open() ;
  
  {
    Context_t* ctx = Context_Create(argc,argv) ;
    
    Bil_GetContext(bil) = ctx ;
  }
    
  Session_Close() ;
  
  return(bil) ;
}



void (Bil_Delete)(void* self)
{
  Bil_t* bil = (Bil_t*) self ;
  
  {
    Context_t* ctx = Bil_GetContext(bil) ;
    
    if(ctx) {
      Context_Delete(ctx) ;
      free(ctx) ;
      Bil_GetContext(bil) = NULL ;
    }
  }
}



int (Bil_Execute)(Bil_t* bil)
{
  int val = 0 ;
  
  Session_Open() ;
  
  {
    val = Exception_SaveEnvironment ;
    
    if(!val) {
      /* Command-line interface */
      Bil_CLI(bil) ;
      /* Graphical user interface */
      //Bil_GUI(bil) ; /* Not done yet */
    } else {
      Message_Direct("An exception occurs with value %d\n",val) ;
    }
  }
    
  Session_Close() ;
  
  return(val) ;
}




/* 
 * Intern functions 
 */
 
#include "CementSolutionChemistry.h"
#include "HardenedCementChemistry.h"
#include "LogActivityOfWaterInBrine.h"

void Bil_CLI(Bil_t* bil)
/** Command-line interface */
{
  Context_t* ctx = Bil_GetContext(bil) ;

  Message_Info("Started on %s",Message_LaunchDate()) ;
  
  if(Context_IsPrintUsage(ctx)) {
    char** argv = (char**) Context_GetPrintUsage(ctx) ;
    
    Bil_PrintUsage(argv[0]) ;
    return ;
  }
  
  if(Context_IsPrintInfo(ctx)) {
    Bil_PrintInfo() ;
    return ;
  }
  
  if(Context_IsHelpOnline(ctx)) {
    Help_HelpOnline() ;
    return ;
  }
  
  if(Context_IsPrintModel(ctx)) {
    char** argv = (char**) Context_GetPrintModel(ctx) ;
    
    /* The standard requires that argv[argc] be a null pointer */
    if(argv[1]) {
      if(!strncmp(argv[1],"all",strlen(argv[1]))) {
        Models_Print(NULL,stdout) ;
      } else {
        char* codename = argv[1] ;
        Models_Print(codename,stdout) ;
      }
    } else {
      Models_Print(NULL,NULL) ;
    }
    return ;
  }
  
  if(Context_IsPrintModule(ctx)) {
    char** argv = (char**) Context_GetPrintModule(ctx) ;
    
    /* The standard requires that argv[argc] be a null pointer */
    if(argv[1]) {
      if(!strncmp(argv[1],"all",strlen(argv[1]))) {
        Modules_Print(NULL) ;
      } else {
        char* codename = argv[1] ;
        Modules_Print(codename) ;
      }
    } else Modules_Print(NULL) ;
    return ;
  }
  
  if(Context_IsReadOnly(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset = DataSet_Create(filename,options) ;
    
    DataSet_Delete(dataset) ;
    free(dataset) ;
    return ;
  }
  
  if(Context_IsCreateInputFile(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    FILE* str = fopen(filename,"r") ;
    
    if(!str) {
      Help_WriteData(filename) ;
      Message_Info("To start the computation, type bil %s\n",filename) ;
    } else {
      fclose(str) ;
      Message_Info("File %s already exists.\n",filename) ;
      Message_Exit ;
    }
    
    return ;
  }
  
  if(Context_IsGraph(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset =  DataSet_Create(filename,options) ;
    Mesh_t* mesh = DataSet_GetMesh(dataset) ;
    char* method = Options_GetGraphMethod(options) ;
      
    Message_Direct("Graph (method %s)\n",method) ;
    Mesh_WriteGraph(mesh,filename,method) ;
    DataSet_Delete(dataset) ;
    free(dataset) ;
    return ;
  }
  
  if(Context_IsInversePermutation(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset =  DataSet_Create(filename,options) ;
    Mesh_t* mesh = DataSet_GetMesh(dataset) ;
    char method[] = "hsl" ;
      
    Message_Direct("Inverse permutations (method %s)\n",method) ;
    Mesh_WriteInversePermutation(mesh,filename,method) ;
    DataSet_Delete(dataset) ;
    free(dataset) ;
    return ;
  }
  
  if(Context_IsNodalOrdering(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset =  DataSet_Create(filename,options) ;
    Mesh_t* mesh = DataSet_GetMesh(dataset) ;
    char* method = Options_GetNodalOrderingMethod(options) ;
      
    Message_Direct("Nodal ordering (method %s)\n",method) ;
    Mesh_WriteInversePermutation(mesh,filename,method) ;
    DataSet_Delete(dataset) ;
    free(dataset) ;
    return ;
  }
  
  if(Context_IsElementOrdering(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset =  DataSet_Create(filename,options) ;
    Mesh_t* mesh = DataSet_GetMesh(dataset) ;
    char* method = Options_GetElementOrderingMethod(options) ;
      
    Message_Direct("Element ordering (method %s)\n",method) ;
    Mesh_WriteInversePermutation(mesh,filename,method) ;
    DataSet_Delete(dataset) ;
    free(dataset) ;
    return ;
  }
  
  if(Context_IsPostProcessing(ctx)) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset =  DataSet_Create(filename,options) ;
    char* method = Options_GetPostProcessingMethod(options) ;
      
    Message_Direct("Post-processing\n") ;
    
    /* GMSH file formats */
    if(strncmp(method,"Gmsh",4) == 0) {
      PosFilesForGMSH_t* pf4gmsh = PosFilesForGMSH_Create(dataset) ;
      char* fmt = method + 4 ;
      
      /* GMSH Parsed file format */
      if(strncmp(fmt,"ParsedFileFormat",strlen(fmt)) == 0) {
        PosFilesForGMSH_ParsedFileFormat(pf4gmsh) ;
      /* GMSH ASCII file format */
      } else if(strncmp(fmt,"ASCIIFileFormat",strlen(fmt)) == 0) {
        PosFilesForGMSH_ASCIIFileFormat(pf4gmsh) ;
      } else {
        Message_FatalError("Format not available") ;
      }
      
      PosFilesForGMSH_Delete(pf4gmsh) ;
      free(pf4gmsh) ;
      
    } else {
      Message_FatalError("Format not available") ;
    }

    DataSet_Delete(dataset) ;
    free(dataset) ;
    return ;
  }
  
  if(Context_IsMiscellaneous(ctx)) {
    double temp = atof(((char**) Context_GetMiscellaneous(ctx))[1]) ;
    
    {
      HardenedCementChemistry_t* hcc = HardenedCementChemistry_Create() ;
    
      HardenedCementChemistry_SetRoomTemperature(hcc,temp) ;
      HardenedCementChemistry_PrintChemicalConstants(hcc) ;
      HardenedCementChemistry_Delete(hcc) ;
    }
    
    {
      CementSolutionChemistry_t* csc = CementSolutionChemistry_Create() ;
      
      CementSolutionChemistry_GetRoomTemperature(csc) = temp ;
      CementSolutionChemistry_PrintChemicalConstants(csc) ;
      CementSolutionChemistry_Delete(csc) ;
    }
    
    return ;
  }
  
  if(Context_IsTest(ctx)) {
    #if 0
    {
      LogActivityOfWaterInBrine_t* law = LogActivityOfWaterInBrine_Create() ;
      LogActivityOfWaterInBrine_Delete(law) ;
      return ;
    }
    #endif
    #if 0
    {
      char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
      Options_t* options = Context_GetOptions(ctx) ;
      DataSet_t* dataset = DataSet_Create1(filename,options) ;
      
      DataSet_Delete(dataset) ;
      free(dataset) ;
      return ;
    }
    #endif
    
    return ;
  }
  
  if(1) {
    char* filename = ((char**) Context_GetInputFileName(ctx))[0] ;
    Options_t* options = Context_GetOptions(ctx) ;
    DataSet_t* dataset =  DataSet_Create(filename,options) ;
    Module_t* module = DataSet_GetModule(dataset) ;

    #if !Threads_APIis(None)
    {
      int nthreads = Options_NbOfThreads(options) ;
      
      Threads_SetTheNbOfThreads(nthreads) ;
      Message_Direct("Nb of requested threads: %d\n",nthreads) ;
      Message_Direct("Multithreading API: "Utils_STR(Threads_API)"\n") ;
    }
    #endif
  
    Message_Direct("Calculation of %s\n",filename) ;
    Message_Direct("Enter in %s\n",Module_GetCodeNameOfModule(module)) ;
    Module_ComputeProblem(module,dataset) ;
    Message_Direct("End of calculation\n") ;
    
    DataSet_Delete(dataset) ;
    free(dataset) ;
    
    Message_Info("CPU time %g seconds\n",Message_CPUTime()) ;
    return ;
  }
  
  return ;
}




void Bil_PrintUsage(char* path)
{
  Message_Direct("Usage: %s [options] file\n",path) ;
  Message_Direct("\n") ;
  Message_Direct("Options:\n") ;
  
  Message_Direct("  -debug \"input\"       Display the data structure of \"input\".\n") ;
  #if 0
  Message_Direct("                       Available inputs are:\n") ;
  Message_Direct("                       - geom   : geometry\n") ;
  Message_Direct("                       - mesh   : mesh\n") ;
  Message_Direct("                       - mate   : materials\n") ;
  Message_Direct("                       - field  : fields\n") ;
  Message_Direct("                       - init   : initialization\n") ;
  Message_Direct("                       - func   : time functions\n") ;
  Message_Direct("                       - bcond  : boundary conditions\n") ;
  Message_Direct("                       - load   : loads\n") ;
  Message_Direct("                       - poin   : points\n") ;
  Message_Direct("                       - obval  : objective variations\n") ;
  Message_Direct("                       - iter   : iterative process\n") ;
  Message_Direct("                       - time   : time steps\n") ;
  Message_Direct("                       - all    : all the dataset above\n") ;
  Message_Direct("                       - inter  : interpolation functions\n") ;
  Message_Direct("                       - continuity: equations and unknowns at nodes\n") ;
  Message_Direct("                       - matrix : the matrix\n") ;
  Message_Direct("                       - residu : the residu\n") ;
  #endif
  
  Message_Direct("  -graph \"fmt\"         Print the graph in \"file.graph\"\n") ;
  Message_Direct("                       in the format \"fmt\".\n") ;
  #if 0
  Message_Direct("                       Available formats are:\n") ;
  Message_Direct("                       - metis: for METIS (if installed),\n") ;
  Message_Direct("                       - hsl_mc40: for HSL_MC40,\n") ;
  #endif
  
  Message_Direct("  -help                Online help.\n") ;
  Message_Direct("  -info                Display general informations.\n") ;
  
  Message_Direct("  -iperm               Print the inverse permutation of nodes \n") ;
  Message_Direct("                       in \"file.graph.iperm\" by \"HSL_MC40\".\n") ;
  
  Message_Direct("  -eordering \"meth\"    Print the inverse permutation of elements\n") ;
  Message_Direct("                       in \"file.graph.iperm\" by the method \"meth\".\n") ;
  #if 0
  Message_Direct("                       Available methods are:\n") ;
  Message_Direct("                       - hsl_mc43\n") ;
  #endif
  
  Message_Direct("  -nordering \"meth\"    Print the inverse permutation of nodes \n") ;
  Message_Direct("                       in \"file.graph.iperm\" by the method \"meth\".\n") ;
  #if 0
  Message_Direct("                       Available methods are:\n") ;
  Message_Direct("                       - hsl_mc40 (same as -iperm)\n") ;
  #endif
  
  /*
  Message_Direct("  -verbosity \"degre\"   Define the verbosity level:\n") ;
  Message_Direct("                       \"degre\" = 0  : nothing\n") ;
  Message_Direct("                       \"degre\" = 1  : normal (default)\n") ;
  Message_Direct("                       \"degre\" = 2  : extra\n") ;
  */
  Message_Direct("  -level \"I\"           Define the screen output level \"I\".\n") ;
  
  Message_Direct("  -model               Display the available models.\n") ;
  
  Message_Direct("  -model \"mymodel\"     Display an example of material properties\n") ;
  Message_Direct("                       for the model \"mymodel\".\n") ;
  
  Message_Direct("  -modules             Display the available modules.\n") ;
  
  Message_Direct("  -solver \"meth\"       Use a solver defined by the method \"meth\".\n") ;
  #if 0
  Message_Direct("                       Available methods are:\n") ;
  Message_Direct("                       - crout : CROUT method (default),\n") ;
  Message_Direct("                       - slu   : SuperLU (if installed).\n") ;
  Message_Direct("                       - ma38  : HSL-MA38 (if installed).\n") ;
  #endif
  
  Message_Direct("  -post \"fmt\"          Generates the post-processing files \n") ;
  Message_Direct("                       \"file.posI\" in the format \"fmt\".\n") ;
  #if 0
  Message_Direct("                       Available formats are:\n") ;
  Message_Direct("                       - GmshParsed: for GMSH parsed file format.\n") ;
  Message_Direct("                       - GmshASCII: for GMSH ASCII file format.\n") ;
  #endif
  
  Message_Direct("  -readonly            Read \"file\" only.\n") ;
  Message_Direct("  -with \"mod\"          Use the module \"mod\".\n") ;
}



void Bil_PrintInfo(void)
{
  char bil_progname[]  = BIL_PROGNAME ;
  char bil_copyright[] = BIL_COPYRIGHT ;
  char bil_version[]   = "Version           : " BIL_VERSION ;
  char bil_license[]   = "License           : " BIL_SHORT_LICENSE ;
  char bil_os[]        = "Build OS          : " BIL_OS ;
  char bil_date[]      = "Build date        : " BIL_DATE ;
  char bil_host[]      = "Build host        : " BIL_HOST ;
  char bil_packager[]  = "Packager          : " BIL_PACKAGER ;
  char bil_url[]       = "Web site          : " BIL_URL ;
  char bil_email[]     = "Contact           : " BIL_EMAIL ;
  char bil_mtapi[]     = "Multithreading API: " Utils_STR(Threads_API) ;
  char bil_nthreads[]  = "Nb of CPU threads : " ;
  int nthreads = Threads_NbOfLogicalCores ;
  
  Message_Direct("%s\n", bil_progname) ;
  Message_Direct("%s\n", bil_copyright) ;
  Message_Direct("%s\n", bil_version) ;
  Message_Direct("%s\n", bil_license) ;
  Message_Direct("%s\n", bil_os) ;
  Message_Direct("%s\n", bil_date) ;
  Message_Direct("%s\n", bil_host) ;
  Message_Direct("%s\n", bil_packager) ;
  Message_Direct("%s\n", bil_url) ;
  Message_Direct("%s\n", bil_email) ;
  Message_Direct("%s\n", bil_mtapi) ;
  Message_Direct("%s%d\n",bil_nthreads,nthreads) ;
}
